src/validation.cpp

```diff
@@ -3754,14 +3753,34 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
             m_blockman.m_dirty_blockindex.insert(to_mark_failed);
         }
 
-        // Add any equal or more work headers to setBlockIndexCandidates
-        auto candidate_it = candidate_blocks_by_work.lower_bound(invalid_walk_tip->pprev->nChainWork);
-        while (candidate_it != candidate_blocks_by_work.end()) {
-            if (!CBlockIndexWorkComparator()(candidate_it->second, invalid_walk_tip->pprev)) {
+        // Mark descendants of the invalidated block as invalid
+        // (possibly replacing a pre-existing BLOCK_FAILED_VALID with BLOCK_FAILED_CHILD)
+        // Add any equal or more work headers that are not invalidated to setBlockIndexCandidates
+        auto candidate_it = highpow_outofchain_headers.lower_bound(invalid_walk_tip->pprev->nChainWork);
+        while (candidate_it != highpow_outofchain_headers.end()) {
+            if (candidate_it->second->GetAncestor(invalid_walk_tip->nHeight) == invalid_walk_tip) {
+                candidate_it->second->nStatus = (candidate_it->second->nStatus & ~BLOCK_FAILED_VALID) | BLOCK_FAILED_CHILD;
+                // If invalidated, the block is irrelevant for setBlockIndexCandidates and m_best_header and can be removed fro mthe cache
+                candidate_it = highpow_outofchain_headers.erase(candidate_it);
+                continue;
+            }
+            if (!CBlockIndexWorkComparator()(candidate_it->second, invalid_walk_tip->pprev) &&
+                candidate_it->second->IsValid(BLOCK_VALID_TRANSACTIONS) &&
+                candidate_it->second->HaveNumChainTxs()) {
                 setBlockIndexCandidates.insert(candidate_it->second);
-                candidate_it = candidate_blocks_by_work.erase(candidate_it);
-            } else {
-                ++candidate_it;
+            }
+            ++candidate_it;
+        }
+
+        // Determine new best header if that was the invalidated block or one of its descendants
+        if (m_chainman.m_best_header->nStatus & BLOCK_FAILED_MASK) {
+            m_chainman.m_best_header = invalid_walk_tip->pprev; // this block is definitely still valid at this point, but there may be better ones
+            auto best_header_it = highpow_outofchain_headers.lower_bound(m_chainman.m_best_header->nChainWork);
+            while (best_header_it != highpow_outofchain_headers.end()) {
+                if (!CBlockIndexWorkComparator()(best_header_it->second, m_chainman.m_best_header)) {
+                    m_chainman.m_best_header = best_header_it->second;
+                }
+                ++best_header_it;
             }
         }
```

Comment:
Can the `m_best_header` update logic be merged into the while loop that handles the `nStatus` updates? Not sure if it is necessarily better; but I was thinking of something like:

Adding this just before the while loop:
```c++
const bool best_header_needs_update = m_chainman.m_best_header->GetAncestor(invalid_walk_tip->nHeight) == invalid_walk_tip;
if (best_header_needs_update) {
  m_chainman.m_best_header = invalid_walk_tip->pprev;
}
``` 

And then here:
```suggestion
            if (best_header_needs_update &&
                !CBlockIndexWorkComparator()(candidate_it->second, m_chainman.m_best_header)) {
                m_chainman.m_best_header = candidate_it->second;
            }
            ++candidate_it;
        }
```

Reply:
Yes, I like that solution, it's less code and since we would `continue`  when encountering an invalid header before we'd never set an invalid block as best header, so it's correct. 
I added you as co-author of that commit with the email from your github profile - let me know if you want met to change anything about that!

Reply:
Thanks a lot! Yes the email is okay.